src/server/game/Movement/MotionMaster.cpp

```diff
@@ -190,27 +191,51 @@ void MotionMaster::Update(uint32 diff)
         ASSERT(result, "MotionMaster:Update: an error occurred during Top removal (%s)", _owner->GetGUID().ToString().c_str());
     }
     else
-        RemoveCleanFlag(MOTIONMMASTER_CLEANFLAG_UPDATE);
+        RemoveFlag(MOTIONMMASTER_FLAG_UPDATE);
 
-    if (HasCleanFlag(MOTIONMMASTER_CLEANFLAG_CLEAR_PENDING))
+    while (!_delayedActions.empty())
     {
-        DirectClear(HasCleanFlag(MOTIONMMASTER_CLEANFLAG_RESET_DEFAULT));
-        RemoveCleanFlag(MOTIONMMASTER_CLEANFLAG_CLEAR_PENDING);
+        _delayedActions.front().Resolve();
+        _delayedActions.pop_front();
     }
 }
 
+void MotionMaster::Add(MovementGenerator* movement, MovementSlot slot)
+{
+/*
+    if (MovementGenerator* curr = _slot[slot])
+    {
+        _slot[slot] = nullptr; // in case a new one is generated in this slot during directdelete
+        if (_top == slot && (_cleanFlag & MOTIONMMASTER_CLEANFLAG_UPDATE))
+            DelayedDelete(curr);
+        else
+            DirectDelete(curr);
+    }
+    else if (_top < slot)
+    {
+        _top = slot;
+    }
+
+    _slot[slot] = m;
+    if (_top > slot)
+        _initialize[slot] = true;
+    else
+    {
+        _initialize[slot] = false;
+        m->Initialize(_owner);
+    }
+*/
+}
+
 bool MotionMaster::Remove(MovementGeneratorType type, MovementSlot slot)
 {
     return false;
 }
 
 void MotionMaster::Clear(bool reset/*= true*/)
 {
-    if (HasCleanFlag(MOTIONMMASTER_CLEANFLAG_UPDATE))
-    {
-        reset ? AddCleanFlag(MOTIONMMASTER_CLEANFLAG_RESET_DEFAULT) : RemoveCleanFlag(MOTIONMMASTER_CLEANFLAG_RESET_DEFAULT);
-        AddCleanFlag(MOTIONMMASTER_CLEANFLAG_CLEAR_PENDING);
-    }
+    if (HasFlag(MOTIONMMASTER_FLAG_UPDATE))
+        _delayedActions.push_back(MotionMasterDelayedClear(this, reset));
```

Comment:
you are slicing the object here - will result in calling compiler generated "pure virtual" (basically, std::terminate)

Reply:
ouch, somehow I missed it >.<